docs/man/man4/scsi.4
.TH SCSI 4
.CT 2 sa
.SH NAME
scsi \- SCSI interface
.SH SYNOPSIS
.B "#include <scsi.h>"
.SH DESCRIPTION
The special file
.F /dev/scsi
provides a low level interface to a SCSI bus.
Commands are transmitted to the bus by
.IR write ;
the response to each command is received with
.IR read (2).
The format of a command is
.IP
.EX
.ta \w'unsigned char scsistatus;  'u
unsigned long tran_id;	/* transaction id */
unsigned char target;	/* SCSI id of target device */
unsigned char flags;	/* flags for this transaction */
unsigned long nreturn;	/* number of bytes to be read back */
char cmd[10];	/* SCSI command */
char data[];	/* optional data */
.EE
.LP
Thus, to send 
.I n
bytes of data, the byte count for
.I write
should be
.IR n +20.
Possible flags are
.B SCSI_WR
(data goes from host to SCSI),
.B SCSI_RD
(data goes from SCSI to host),
.B SCSI_BRESET
(reset the SCSI bus),
.B SCSI_RESET
(reset the controller),
and
.B SCSI_SENSE
(return extended sense data on error).
For most controllers, 
.B SCSI_BRESET
implies
.BR SCSI_RESET .
Flags are OR'ed together, and there must be exactly one of
.B SCSI_WR
and
.BR SCSI_RD .
The SCSI command should terminate within a small time
(currently 10 seconds);
a longer limit (300 seconds) can be specified by using
.BR SCSI_LTMOUT .
.LP
The data read is structured as
.IP
.EX
unsigned long tran_id;	/* transaction id */
unsigned char status;	/* scsi status byte */
unsigned char message;	/* scsi message byte */
unsigned char flags;	/* flags for this transaction */
unsigned char c_type;	/* 1=td 2=us */
unsigned short c_reg1;	/* td=sa, us=csr */
unsigned short c_reg2;	/* td=mscp, us=per */
unsigned char sense[22];	/* extended sense data */
char pad[2];
char data[];	/* any data */
.EE
.LP
Thus, to read 
.I n
bytes of data. the byte count to
.I read
should be
.IR n +36.
If
.I flags
has the
.B SCSI_CERR
bit set,
there was a controller error, which is described by the
.B c_
fields.
The values of
.B csr
(or
.BR sa )
and
.B per
(or
.BR mscp )
are documented in the interface manual for the U.S. Design 1158
Unibus controller (or T.D. Systems Viking controller).
If the
.B SCSI_SENSE
bit was set in the 
.I write,
and the status byte shows a check condition,
an attempt is made to get extended sense information.
If the attempt succeeds the
.B SCSI_SENSE
is set in
.I flags.
Otherwise, the status and message bytes for the failed
attempt are placed in
.B sense[0]
and
.B sense[1]
respectively.
.PP
The transaction id identifies which
.I write
caused the results for this
.IR read .
This will become more important when multiple simultaneous transactions
are allowed.
.SH FILES
.F /dev/scsi
.SH "SEE ALSO"
.IR scsish (8)


v10/sys/io/scsi.c
/*
	SCSI Pass-Thru driver for the TD Systems UD? -- Andrew Hume
	Ninth Edition Unix
*/

#include "sys/param.h"
#include "sys/user.h"
#include "sys/buf.h"
#include "sys/systm.h"
#include "sys/pte.h"
#include "sys/map.h"
#include "sys/ubaddr.h"
#include "sys/conf.h"

#include	"sys/uda.h"
#include	"sys/mscp.h"

#include "sys/scsi.h"
#include	<scsi.h>

#define	mscp_scsi	m_fmt

struct udadevice {
	short	udaip;		/* initialization and polling */
	short	udasa;		/* status and address */
};
#define	UDA_ERR		0100000	/* error bit */
#define	UDA_STEP4	0040000	/* step 4 has started */
#define	UDA_STEP3	0020000	/* step 3 has started */
#define	UDA_STEP2	0010000	/* step 2 has started */
#define	UDA_STEP1	0004000	/* step 1 has started */
#define	UDA_GO		0000001	/* start operation, after init */

#define	UDA_OWN		0x80000000	/* UDA owns this descriptor */
#define	UDA_INT		0x40000000	/* allow interrupt on ring transition */

#define WAITPRI 	(PZERO+1)
#define	SCSILTMT	300		/* long timeout in seconds */
#define	SCSISTMT	10		/* short timeout in seconds */
#define	ERROR		0x8000		/* in csr */
#define	SCSIINTR	0700
#define	SETDSC(dsc, fl)	dsc = (dsc&~(UDA_OWN|UDA_INT))|fl

static memset(p, c, n)
	register char *p, c;
	register n;
{
	while(n-- > 0)
		*p++ = c;
}

int scsiopen(), scsiclose(), scsiread(), scsiwrite();

extern struct scsi scsi[];
extern struct ubaddr scsiaddr[];
extern int scsicnt;
struct cdevsw scsicdev =
	cdinit(scsiopen, scsiclose, scsiread, scsiwrite, nodev);
int scsi_dbg = 0;

scsiopen(dev)
	dev_t dev;
{
	register struct scsi *p;

	if((dev = minor(dev)) >= scsicnt) {
		u.u_error = ENODEV;
		return;
	}
	if((p = &scsi[dev])->flag&OPEN) {
		u.u_error = EBUSY;
		return;
	}
	if(((p->addr = (struct udadevice *)ubaddr(&scsiaddr[dev])) == 0)
	  || ubbadaddr(scsiaddr[dev].ubno, (caddr_t)p->addr, sizeof(u_short))) {
		printf("scsi%d absent\n", dev);
		u.u_error = ENODEV;
		return;
	}
	if((p->flag&USED) == 0)
		if(scsistart(dev))
			return;
	p->flag = USED|OPEN|NEXTWR|DONE;
}

scsistart(dev)
	dev_t dev;
{
	register struct scsi *p = &scsi[minor(dev)];

	p->flag |= DONE;
	p->b1 = geteblk();
	p->b1->b_bcount = BUFSIZE;
	p->b1->b_flags = B_BUSY;
	clrbuf(p->b1);
	p->ub1 = ubmbuf(scsiaddr[minor(dev)].ubno, p->b1, USLP);
	p->u1 = ubadbuf(scsiaddr[minor(dev)].ubno, p->b1, p->ub1);
	p->data = (unsigned char *)p->b1->b_un.b_addr;
	p->b2 = geteblk();
	p->b2->b_bcount = BUFSIZE;
	p->b2->b_flags = B_BUSY;
	clrbuf(p->b2);
	p->ub2 = ubmbuf(scsiaddr[minor(dev)].ubno, p->b2, USLP);
	p->u2 = ubadbuf(scsiaddr[minor(dev)].ubno, p->b2, p->ub2);
	p->junk = (struct bag *)p->b2->b_un.b_addr;
	p->timeout = SCSISTMT;
	if(scsiinit(dev) == 0){
		scsiclose(dev);
		p->flag = 0;
		u.u_error = ENXIO;
		return(1);
	}
	printf("scsi%d run\n", minor(dev));
	return(0);
}

scsiclose(dev)
	dev_t dev;
{
	register struct scsi *p = &scsi[minor(dev)];

	if((p->flag&DONE) == 0){	/* wait for I/O to complete */
		(void)tsleep((caddr_t)p, PZERO+1, p->timeout);
	}
	p->flag = USED;
}

scsiwrite(dev)
	dev_t dev;
{
	register count;
	register struct scsi *p = &scsi[minor(dev)];
	register struct mscmd *cmd = &p->junk->cmd.msg;
	unsigned char flag, bus_id;

	if(p->flag&NEXTWR)
		p->flag &= ~NEXTWR;
	else {
		u.u_error = EGREG;
		return;
	}
	if(copyin(u.u_base, &p->tran_id, 4) || copyin(u.u_base+4, &bus_id, 1)
			|| copyin(u.u_base+5, &flag, 1)
			|| copyin(u.u_base+6, &cmd->mscp_scsi, SCSICMD)){
		u.u_error = EFAULT;
		return;
	}
	if(flag & SCSI_RESET){
		if(scsiinit(dev) == 0){
			printf("scsi%d: reset failed\n", minor(dev));
			p->flag = 0;
			u.u_error = ENXIO;
			return;
		}
		printf("scsi%d: reset\n", minor(dev));
		p->flag = USED|OPEN|NEXTWR|DONE;
		u.u_count = 0;
		return;
	}
	p->timeout = (flag&SCSI_LTMOUT)? SCSILTMT:SCSISTMT;
	count = u.u_count - (6+SCSICMD);
	u.u_base += 6+SCSICMD;
	if((count < 0) || (count > SCSIDATA)){
		u.u_error = EINVAL;
		return;
	}
	memset(p->data, 0xEE, BUFSIZE);
	if(flag&SCSI_RD){
		cmd->m__r1 = 0106;
		count = scsilen((unsigned char *)&cmd->mscp_scsi);
	} else if(flag&SCSI_WR){
		if(copyin(u.u_base, p->data, count)){
			u.u_error = EFAULT;
			return;
		}
		cmd->m__r1 = 0107;	
	} else {
		u.u_error = EINVAL;
		return;
	}
	cmd->m_opcd = (flag & SCSI_BRESET)? 0160 : 0130;
	if(scsi_dbg)
		printf("c=%d uc=%d, dir=0%o, bus_id=%d, new count=%d timeout=%d\n",
			count, u.u_count, cmd->m__r1, bus_id, count, p->timeout);
	cmd->m_unit = bus_id;
	cmd->m_bcnt = count;
	cmd->m_fcnt = p->u1;
	p->flag = (p->flag&~DONE)|PEND;
	if(flag&SCSI_SENSE)
		p->flag |= ERRSNSE;
	p->junk->ca.ca_cmdint = 0;
	p->junk->ca.ca_rspint = 0;
	SETDSC(p->junk->ca.ca_cmddsc[0], UDA_OWN);
	bus_id = p->addr->udaip;	/* start controller */
	u.u_count = 0;
}

scsiread(dev)
	dev_t dev;
{
	register struct scsi *p = &scsi[minor(dev)];
	register count;

	if(p->flag&NEXTWR){
		u.u_error = EGREG;
		return;
	} else
		p->flag |= NEXTWR;
	if((p->flag&DONE) == 0){
		if(tsleep((caddr_t)p, PZERO+1, p->timeout) != TS_OK){
			u.u_error = ENXIO;
			return;
		}
	}
	if(p->sa&0x8000){
		printf("scsi%d: error sa=#%x\n", minor(dev), p->sa&0xFFFF);
		u.u_error = EIO;
		scsiinit(dev);
		return;
	}
	count = u.u_count - SCSIRET;
	if(count > p->junk->rsp.msg.m_bcnt)
		count = p->junk->rsp.msg.m_bcnt;
	if((count < 0) || (count > SCSIDATA)){
		u.u_error = EINVAL;
		return;
	}
	((unsigned long *)p->status)[0] = p->tran_id;
	p->status[4] = p->junk->rsp.msg.m_fbbk;
	p->status[5] = p->junk->rsp.msg.m_fbbk>>8;
	p->status[6] = 0;	/* flags */
	p->status[7] = 1;	/* viking */
	((short *)p->status)[4] = p->sa;
	((short *)p->status)[5] = p->junk->rsp.msg.m_sts;
	if(copyout(p->status, u.u_base, SCSIRET)){
		u.u_error = EFAULT;
		return;
	}
	if(count)
		if(copyout(p->data, u.u_base+SCSIRET, count)){
			u.u_error = EFAULT;
			return;
		}
	if(p->junk->rsp.msg.m_sts){
		printf("scsi%d: stat=#%x\n", minor(dev), p->junk->rsp.msg.m_sts);
		u.u_error = (p->junk->rsp.msg.m_sts == 0x80a)? ESRCH : EIO;
		return;
	}
	u.u_count -= SCSIRET+count;
	SETDSC(p->junk->ca.ca_rspdsc[0], UDA_OWN|UDA_INT);
}

scsi0int(dev)
	dev_t dev;
{
	register struct scsi *p = &scsi[minor(dev)];
	register s;

	if((p->flag&PEND) == 0){
		printf("scsi%d: unexpected interrupt\n", minor(dev));
		return;
	}
	if(p->junk->ca.ca_rspint == 0){
		printf("scsi%d: rspint=0 cmdint=%d rsp=#%x cmd=#%x\n", minor(dev), p->junk->ca.ca_cmdint, p->junk->ca.ca_rspdsc[0], p->junk->ca.ca_cmddsc[0]);
		return;
	}
	p->sa = p->addr->udasa;
	s = spl6();
	p->flag = (p->flag&~PEND)|DONE;
	splx(s);
	wakeup((caddr_t)p);
}

/*
 * hardware initialization handshake
 * for simplicity, don't bother with init interrupts
 * returns nonzero if ok
 */

#define	UDA_STEPS	(UDA_ERR|UDA_STEP4|UDA_STEP3|UDA_STEP2|UDA_STEP1)

scsiinit(d)
int d;
{
	register struct scsi *p = &scsi[minor(d)];
	register struct udadevice *udaddr;
	register int i;
	time_t out;
	int s;

	udaddr = p->addr;
	out = time + 11;
	s = spl0();
printf("scsi1");
	udaddr->udaip = 0;
	while((udaddr->udasa & (UDA_ERR|UDA_STEP1)) == 0 && time <= out)
		;
	if((udaddr->udasa & UDA_STEPS) != UDA_STEP1) {
		steperr("1", udaddr->udasa, d);
		splx(s);
		return (0);
	}
	udaddr->udasa = UDA_ERR|(SCSIINTR/4);
	/* no diagnostic wrap, no interrupts during initialization */
printf("2");
	out = time + 21;
	while((udaddr->udasa & (UDA_ERR|UDA_STEP2)) == 0 && time <= out)
		;
	if((udaddr->udasa & UDA_STEPS) != UDA_STEP2) {
		steperr("2", udaddr->udasa, d);
		splx(s);
		return (0);
	}
	i = p->u2;	/* unibus address of bag */
	i += (long)&p->junk->ca.ca_rspdsc[0] - (long)p->junk;
	udaddr->udasa = (short)i;
	out = time + 11;
printf("3");
	while((udaddr->udasa & (UDA_ERR|UDA_STEP3)) == 0 && time <= out)
		;
	if((udaddr->udasa & UDA_STEPS) != UDA_STEP3) {
		steperr("3", udaddr->udasa, d);
		splx(s);
		return (0);
	}
	udaddr->udasa = (i >> 16)&0x3F;
	out = time + 11;
printf("4");
	while((udaddr->udasa & (UDA_ERR|UDA_STEP4)) == 0 && time <= out)
		;
	if((udaddr->udasa & UDA_STEPS) != UDA_STEP4) {
		steperr("4", udaddr->udasa, d);
		splx(s);
		return (0);
	}
	i = p->u2 + (long)&p->junk->rsp.msg.m_crf - (long)p->junk;
	p->junk->ca.ca_rspdsc[0] = UDA_OWN|UDA_INT|i;
	i = p->u2 + (long)&p->junk->cmd.msg.m_crf - (long)p->junk;
	p->junk->ca.ca_cmddsc[0] = i;
	if (udaddr->udasa & UDA_ERR) {
		steperr("5", udaddr->udasa, d);
		splx(s);
		return (0);
	}
	udaddr->udasa = UDA_GO;		/* finish init */
	/* finish cmd packet init */
	p->junk->cmd.msg_len = 44;
	p->junk->cmd.msg.m_crf = (long)p->junk;
	p->junk->cmd.msg.m_opcd = 0130;
	splx(s);
printf(" done\n");
	return (1);
}

steperr(str, sa, dev)
	char *str;
{
	printf("scsi%d: step%s error, sa=#%x\n", minor(dev), str, sa&0xFFFF);
}

/*
	dreck to figure out how much we should read back
*/

scsilen(cmd)
	unsigned char *cmd;
{
	switch(cmd[0])
	{
	case 0x03:	return(cmd[4]);
	case 0x08:	return(cmd[4]*1024);
	case 0x12:	return(cmd[4]);
	case 0x1A:	return(cmd[4]);
	case 0x1C:	return(cmd[3]*256 + cmd[4]);
	case 0x25:	return(8);
	case 0x28:	return(1024*(256*cmd[7] + cmd[8]));
	case 0x2C:	return(6);
	case 0x2D:	return(6);
	case 0x4D:	return(1*(256*cmd[7] + cmd[8]));
	case 0xC2:	return(1024);
	case 0xC3:	return(4096);
	case 0xD3:	return(20);
	}
	return(0);
}
